meta: add some clarification to the nomination process (#57503)
nodejs/node: GOVERNANCE.md

* Participation in other projects, teams, and working groups of the Node.js
  organization.

Collaborators should be people volunteering to do unglamorous work because it's
the right thing to do, they find the work itself satisfying, and they care about
Node.js and its users. People should get collaborator status because they're
doing work and are likely to continue doing work where having the abilities that
come with collaborator status are helpful (abilities like starting CI jobs,
reviewing and approving PRs, etc). That will usually--but, very importantly, not
always--be work involving commiting to the `nodejs/node` repository. For an example
of an exception, someone working primarily on the website might benefit from being
able to start Jenkins CI jobs to test changes to documentation tooling. That,
along with signals indicating commitment to Node.js, personal integrity, etc.,

> should be enough for a successful nomination.

Reviewer: This is great.

Author: These are @Trott's words :-)

  organization.

Collaborators should be people volunteering to do unglamorous work because it's
the right thing to do, they find the work itself satisfying, and they care about
Node.js and its users. People should get collaborator status because they're
doing work and are likely to continue doing work where having the abilities that
come with collaborator status are helpful (abilities like starting CI jobs,
reviewing and approving PRs, etc). That will usually--but, very importantly, not
always--be work involving commiting to the `nodejs/node` repository. For an example
of an exception, someone working primarily on the website might benefit from being
able to start Jenkins CI jobs to test changes to documentation tooling. That,
along with signals indicating commitment to Node.js, personal integrity, etc.,
should be enough for a successful nomination.

It is important to understand that potential collaborators may have vastly
different areas and levels of expertise, interest, and skill. The Node.js
project is large and complex, and it is not expected that every collaborator
will have the same level of expertise in every area of the project. The

> complexity or "sophistication" of an individuals contributions, or even their

Reviewer 2: ```suggestion
complexity or "sophistication" of an individual’s contributions, or even their
```

* Participation in other projects, teams, and working groups of the Node.js
  organization.

Collaborators should be people volunteering to do unglamorous work because it's
the right thing to do, they find the work itself satisfying, and they care about

> Node.js and its users. People should get collaborator status because they're

Reviewer 3: @jasnell could you emphasize then that a collaborator is not just someone working on nodejs/node

Reviewer 4: We list what contributions are expected to be relevant: https://github.com/nodejs/node/blob/df883b0c7234cb44bef3175cd3e6a997c830db05/GOVERNANCE.md#L173-L190

Does that address your concern?

Reviewer 3: I don't think so, this is what we had already. Maybe just using **bold** around those lines is enough. But it is fine if y'all believe leave as it is.

Author: I lean towards keeping this as is but can be convinced. I think such changes could come in a second PR tho.

Reviewer 3: Fine by me!

doing work and are likely to continue doing work where having the abilities that
come with collaborator status are helpful (abilities like starting CI jobs,

> reviewing and approving PRs, etc). That will usually--but, very importantly, not

Reviewer 5: ```suggestion
reviewing and approving PRs, etc.). That will usually--but, very importantly, not
```

> always--be work involving commiting to the `nodejs/node` repository. For an example

Reviewer 5: ```suggestion
always--be work involving committing to the `nodejs/node` repository. For an example
```

of an exception, someone working primarily on the website might benefit from being
able to start Jenkins CI jobs to test changes to documentation tooling. That,
along with signals indicating commitment to Node.js, personal integrity, etc.,
should be enough for a successful nomination.

> It is important to understand that potential collaborators may have vastly

Reviewer 3: @jasnell can you please add a section describing examples of what a potential collaborator candidate would be? What are contributions we should look for? Creating a baseline will help a lot of people to not feel uncomfortable doing so!

different areas and levels of expertise, interest, and skill. The Node.js
project is large and complex, and it is not expected that every collaborator
will have the same level of expertise in every area of the project. The
complexity or "sophistication" of an individual’s contributions, or even their
relative engineering "skill" level, are not primary factors in determining
whether they should be a collaborator. The primary factors do include the quality
of their contributions (do the contributions make sense, do they add value, do
they follow documented guidelines, are they authentic and well-intentioned,

> etc), their commitment to the project, can their judgement be trusted, and do

Reviewer 5: ```suggestion
etc.), their commitment to the project, can their judgement be trusted, and do
```

>    block, add a comment in the private discussion stating you're planning on
>    opening a public issue, e.g. "I see a number of approvals and no block, I'll
>    be opening a public nomination issue if I don't hear any objections in the

Reviewer 4: ```suggestion
   block, and any questions/concerns have been addressed, add a comment in the
   private discussion stating you're planning on opening a public issue, e.g.
   "I see a number of approvals and no block, I'll be opening a public
   nomination issue if I don't hear any objections in the next 72 hours".
```

   next 72 hours". _Ideally_ this should only be done after any questions,

>    concerns, objects, etc have been resolved.

Reviewer 4: For the same reason navigation apps will tell you "After 500 meters, take a right" and not "Take a right after 500 meters", we should change the order here (otherwise, folks might first take the action before reading that they shouldn't have). So it should be something like "…if the nomination proposal has received some support and no explicit block, and any questions/concerns have been addressed, add a comment…"

Author: To make sure the edit is correct can I ask you to possibly use the suggestion mechanism.

Reviewer 4: It's too far up for me to do that unfortunately, GH won't let me

Mention @nodejs/collaborators in the issue to notify other collaborators about
the nomination.

The nomination passes if no collaborators explicitly oppose it after one week.
In the case of an objection, the TSC is responsible for working with the
individuals involved and finding a resolution. The TSC may, following
typical TSC consensus seeking processes, choose to advance or block a
nomination that has otherwise failed to reach a natural consensus or clear
path forward even if there are outstanding objections.

Explicit opposition would typically be signaled as some form of clear
and unambiguous comment like, "I don't believe this nomination should pass"

> or use of the thumbs-down emoji reaction. _Ideally_ these would be paired

Reviewer 4: nit: I don't think the thumbs-down recommendation is a good idea, it doesn't send a notification, it would be hard to track down, and offers no nuance. I'd the explicit comment to be required
```suggestion
. _Ideally_ this would be paired
```

> or use of the thumbs-down emoji reaction. _Ideally_ these would be paired

Reviewer 6: Here I would say "these must be paired".

Reviewer 7: If it is "must" instead of "would" or "should", then "Ideally" also needs to be removed.

Author: I don't think "must" is necessary here so long as it is clear that the objection can be overridden if the strong recommendation is not followed.

4. Open an issue in the [nodejs/node][] repository. Provide a summary of
   the nominee's contributions (see below for an example). Mention
   @nodejs/collaborators in the issue to notify other collaborators about
   the nomination. _Ideally_, this issue should not be opened until any private

>    discussion (questions, concerns, objections, etc) has been resolved.

Reviewer 4: Is this addition necessary given the suggested comment already says that? In any case, I would recommend putting it before "open an issue", otherwise they might already have opened the issue when reading this sentence.

Collaborators might overlook someone with valuable contributions. In that case,
the contributor may open an issue or contact a collaborator to request a

> nomination. Such "self-nominations" follow the same process as regular
> nomination -- that is, the contributor is not nominated until an existing
> Collaborator formally proposes it and opens the discussion.

Reviewer 4: We don't have consensus over this one, I don't think it makes sense to make changes to the self-nomination, AFAIK it has never been a problem, I don't see a reason to make changes.
```suggestion
nomination.
```

Reviewer 7: I'd support removing mention of self-nomination entirely. Anyone with sufficient involvement in Node.js (where becoming a collaborator would benefit them and/or Node.js) is going to have plenty of names (or at least GitHub handles and email addresses and Slack handles) of people to ask. There is never a need to self-nominate. The self-nomination process has been entirely superfluous the entire time I've been involved in Node.js. Simplify things and get rid of it.

Reviewer 7: > I'd support removing mention of self-nomination entirely. Anyone with sufficient involvement in Node.js (where becoming a collaborator would benefit them and/or Node.js) is going to have plenty of names (or at least GitHub handles and email addresses and Slack handles) of people to ask. There is never a need to self-nominate. The self-nomination process has been entirely superfluous the entire time I've been involved in Node.js. Simplify things and get rid of it.

https://github.com/nodejs/node/pull/57537

  organization
* Other participation in the wider Node.js community

> The nomination passes if no collaborators _explicitly_ oppose it after one week.

Reviewer 4: I don't think "explicitly" helps, AFAICT the thing that needs to be clarified is whether such opposition needs to be public to count. Since we're defining below what a "valid" opposition is, let's refer to that instead.
```suggestion
The nomination passes if no collaborators oppose it (as described in the following section) after one week
```

has otherwise failed to reach a natural consensus or clear path forward even
if there are outstanding objections.

Explicit opposition would typically be signaled as some form of clear
and unambiguous comment like, "I don't believe this nomination should pass".
These _should_ be paired with clear suggestions for positive, concrete,
and unambiguous next steps that the nominee can take to overcome the objection
and allow it to pass. While such suggestions are technically optional, they are
_strongly encouraged_ to prevent the nomination from stalling indefinitely or
objections from being overridden by the TSC.

An important rule of thumb is that the nomination process is intended to be
biased strongly towards implicit approval of the nomination. This means
discussion and review around the proposal should ideally be more geared
towards "I have reasons to say no..." as opposed to "Give me reasons to say
yes...".

Refrain from discussing or debating aspects of the nomination process
itself directly within a nomination private discussion or public issue.
Such discussions can derail and frustrate the nomination and cause unnecessary

> friction. Move such discussions to a separate issue or discussion thread.

Reviewer 4: Should this be moved to `How to review a collaborator nomination` section? Or maybe we need a new section `How to oppose to a collaborator nomination`

be cases where collaborators disagree on whether a nomination should move
forward.

An important rule of thumb is that the nomination process is intended to be
biased strongly towards implicit approval of the nomination. This means
discussion and review around the proposal should ideally be more geared
towards "I have reasons to say no..." as opposed to "Give me reasons to say
yes...".

When concerns have been raised on the private discussion, refrain from opening
the public issue. If no one has explicitly blocked the nomination and you'd like
it to move forward, comment something like "If I don't hear any objections
before (some time), I will open the public issue".

Also refrain from discussing or debating aspects of the nomination process
itself directly within a nomination private discussion or public issue.
Such discussions can derail and frustrate the nomination and cause unnecessary

> friction. Move such discussions to a separate issue or discussion thread.

Reviewer 4: Just a heads-up  that this is being moved to the previous section in https://github.com/nodejs/node/pull/57483

* If you are neutral, or feel you don't know enough to have an informed opinion,
  it's certainly OK to not interact with the nomination.
* If you think the nomination was made too soon, or can be detrimental to the
  project, share your concerns. See the section "How to oppose a collaborator
  nomination" below.

Our goal is to keep gate-keeping at a minimal, but it cannot be zero since being
a collaborator requires trust (collaborators can start CI jobs, use their veto,
push commits, etc.), so what's the minimal amount is subjective, and there will
be cases where collaborators disagree on whether a nomination should move
forward.

> Refrain from discussing or debating aspects of the nomination process

Reviewer 3: <3

itself directly within a nomination private discussion or public issue.
Such discussions can derail and frustrate the nomination causing unnecessary
friction. Move such discussions to a separate issue or discussion thread.

##### How to oppose a collaborator nomination

An important rule of thumb is that the nomination process is intended to be
biased strongly towards implicit approval of the nomination. This means
discussion and review around the proposal should be more geared towards "I have
reasons to say no..." as opposed to "Give me reasons to say yes...".

Given that there is no "Request for changes" feature in discussions and issues,
try to be explicit when your comment is expressing a blocking concern.
Similarly, once the blocking concern has been addressed, explicitly say so.

Explicit opposition would typically be signaled as some form of clear
and unambiguous comment like, "I don't believe this nomination should pass".
Asking clarifying questions or expressing general concerns is not the same as

> explicitly opposition; however, a best effort should be made to answer such

Author: ```suggestion
explicit opposition; however, a best effort should be made to answer such
```

be cases where collaborators disagree on whether a nomination should move
forward.

Refrain from discussing or debating aspects of the nomination process
itself directly within a nomination private discussion or public issue.

> Such discussions can derail and frustrate the nomination and cause unnecessary

Reviewer 4: feel free to disregard if that's a valid form of written English I'm not familiar with
```suggestion
Such discussions can derail and frustrate the nominator and cause unnecessary
```

Reviewer 8: The original is probably more the intent because it implies that everyone involved gets frustrated, not just the nominator. The “frustration” is redundant with “friction“ on the next line.

```suggestion
Such discussions can derail the nomination and cause unnecessary
```

friction. Move such discussions to a separate issue or discussion thread.

##### How to oppose a collaborator nomination

> Explicit opposition would typically be signaled as some form of clear
> and unambiguous comment like, "I don't believe this nomination should pass".

Reviewer 4: IMO it's worth point out that opposition doesn't need to be public, and can be anonymous.
```suggestion
Explicit opposition would typically be signaled as some form of clear
and unambiguous comment like, "I don't believe this nomination should pass".
It doesn't need to be done publicly, but the TSC is more likely to dismiss an
opposition done anonymously. Ideally, the comment showing opposition should
be done in the private discussion, ideally before the public issue is opened.
```

We should probably change the bullet point on line 245 "If you think the nomination was made too soon, …" as it repeats a lot of the content of this section, it should instead refer to it.

be cases where collaborators disagree on whether a nomination should move
forward.

> #### The Authenticity of Contributors

Author: It feels like a subsection of the "Nominating a New Contributor" section (which it is currently) and not of the `How to review a collaborator nomination" section... so I don't think this suggestion makes sense.

> #### The Authenticity of Contributors

Reviewer 4: Maybe it should be in `Ideal Nominees` then?
